### ISSUE TITLE
FID score of a dataset to itself does not give 0

### ISSUE THREAD

**Issue by ogencoglu** [mseitzer/pytorch-fid#26]
By definition, the FID score has a minimum bound of 0 but I got -4.8e-5 when I calculated the distance of a dataset to itself. Obviously, this is very close to zero. Nevertheless, can't think of a reason why would this be non-zero from your implementation. Any ideas?

**Comment by mseitzer** [comment]
It might be the case that the list of images in both cases is not ordered in the same way, thus giving slight differences when computing feature mean and covariance.

You can try to insert a `sorted(...)` at this line to avoid this problem:
https://github.com/mseitzer/pytorch-fid/blob/f9cdfd32af6bbb3ed0a484746b401f085457d9a8/fid_score.py#L219

Another possibility is that certain Pytorch functions are non-deterministic on the GPU (see https://pytorch.org/docs/stable/notes/randomness.html#pytorch) and this can obviously impact results. Were you running on GPU? You can try running on CPU instead.

Alternatively, you can insert the following lines somewhere and try again on GPU (although this does not remove all sources of non-determinism as far as I understood):
```
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

If would be very interested to hear if my proposed changes result in different numbers. I could then make the appropriate changes to the code.

**Comment by ogencoglu** [comment]
Yes, I am using GPU. 

PyTorch and other frameworks are non-deterministic during _train time_, they can also act non-deterministic on inference time _for different environments/GPUs_ but it does not make sense that a model is non-deterministic during **inference time on the same environment**. This is the situation here.

**Comment by mseitzer** [comment]
Do you have a reference for CUDA/cuDNN behaving different between training and inference? If some operations are implemented in a non-deterministic way (this is the case), and those operations are used during inference, surely inference can be non-deterministic too.

**Comment by ogencoglu** [comment]
Inception is a basic CNN and inference is just bunch of matrix multiplication. No dropout, no nothing. There is nothing non-deterministic about it. 

CUDA giving slightly different inference/training results across different hardware architectures is due to floating point calculations etc., not because convolution operation is non-deterministic. That's why same device/environment should give the same result in this project.

**Comment by mseitzer** [comment]
Okay, now I see where you are coming from, but I think you may not have the full picture here. The way some operations are implemented in CUDA may involve non-determinism to improve performance, *even if the operations mathematically do not involve randomness* (see https://pytorch.org/docs/stable/notes/randomness.html#pytorch)

Indeed, this even affects convolution, see the note in the documentation:
https://pytorch.org/docs/stable/nn.html#torch.nn.Conv2d

You can disable this behavior by setting
```
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

**Comment by ogencoglu** [comment]
Really interesting. Thanks.

After setting PyTorch determinism to `True` and sorting the lists, I am getting the _same_ result over and over now. It is still a small, non-zero number though. So I believe this is simply a numerical approximation error on the scipy/numpy side.

You can close this now.
